File: ruby/src/detectors/sql-injection/sql-injection.rb
=begin
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
=end

# {fact rule=sql-injection@v1.0 defects=1}
require 'pg'

def sql_injection_noncompliant(event:, context:)
  conn = PG::Connection.open(:dbname => 'test')

  # Noncompliant: User-controlled parameter is used in SQL Statement.
  res2 = conn.exec_params('SELECT * FROM foobar WHERE id = %{id}' % {id: event['id']})
  
end
# {/fact}

# {fact rule=sql-injection@v1.0 defects=0}
require 'pg'

def sql_injection_compliant(event:, context:)
  conn = PG::Connection.open(:dbname => 'test')

  # Compliant: Parameterized SQL Statement.
  res = conn.exec_params('SELECT $1 AS a, $2 AS b, $3 AS c', [event['id'], 2, nil])
  
end
# {/fact}

File: csharp/src/detectors/sql-injection/SQLInjection.cs
using System.Data.SqlClient;

namespace CSharpInjection
{
    class SQLInjection
    {
        // {fact rule=sql-injection@v1.0 defects=1}
        public void SqlInjectionNoncompliant(string sql)
        {
            using (SqlConnection connection = 
                   new SqlConnection("Data Source=(local);Initial Catalog=Northwind;Integrated Security=SSPI;")) {
                connection.Open();
                // Noncompliant: Detected a formatted string in a SQL statement.
                SqlCommand command= new SqlCommand(sql);
            }
        }
        // {/fact}
        
        // {fact rule=sql-injection@v1.0 defects=0}
        public void SqlInjectionCompliant()
        {
            using (SqlConnection connection =
                   new SqlConnection("Data Source=(local);Initial Catalog=Northwind;Integrated Security=SSPI;"))
            {
                string sql = "SELECT * FROM Customers WHERE EmployeeId = @EmployeeId";
                // Compliant: Obtain a PreparedStatement 'SqlParameter'.
                SqlCommand command = new SqlCommand(sql);
                command.Parameters.Add(new SqlParameter("@EmployeeId", System.Data.SqlDbType.Int));
            }
        }
        // {/fact}
    }
}